### PR TITLE
removes last traces of apex etc from PL module

### PR DESCRIPTION
**Pull request by williamFalcon**
removes last traces of apex etc from PL module

### CODE REVIEW

File: pytorch_lightning/core/hooks.py [hunk 1]
```diff
@@ -295,36 +295,23 @@ def on_after_backward(self):
...
-        """
-        loss.backward()
-
-    def amp_scale_loss(self, unscaled_loss: Tensor, optimizer: Optimizer, optimizer_idx: int, amp_backend: AMPType):
-        if amp_backend == AMPType.NATIVE:
-            scaled_loss = self.trainer.scaler.scale(unscaled_loss)
-        else:
-            scaled_loss = amp.scale_loss(unscaled_loss, optimizer)
+                # automatically applies scaling, etc...
+                self.backward(loss)
```
Comment (justusschock) [inline pytorch_lightning/core/hooks.py:311]: While this is certainly an improvement, I'm just worried since one of our main principles was that "it's just organised Pytorch and no new framework to learn". I feel like this really tends towards not being really pytorchic (or how to call that). In PyTorch you call backward on tensors, not on models.

**Review by justusschock — APPROVED** [review]
Looks good to me so far, just one minor concern about design principles. I know, that there probably is no better solution for this, but still it feels like this is not 100% what we're aiming for.

Actually I didn't want to approve(yet) due to failing tests, but pressed the wrong button :D

File: pytorch_lightning/plugins/apex.py [hunk 2]
```diff
@@ -34,6 +34,10 @@ def training_step(self, fx, args):
         output = fx(args)
         return output
 
+    def backward(self, unscaled_loss, optimizer=None, *args, **kwargs):
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/apex.py:37]: Do we want to set optimizer defaults to `None`? Makes sense in native amp as this isn't used, but I think APEX will crash if optimizer is `None`:

https://github.com/NVIDIA/apex/blob/master/apex/amp/handle.py#L17

File: pytorch_lightning/core/hooks.py (hunk 1 above)
Comment (williamFalcon) [reply pytorch_lightning/core/hooks.py:311]: welll... this is optional.

the user can always call loss.backward themselves.

But if they do, they don't get automatic mixed precision.

Comment (williamFalcon) [reply pytorch_lightning/core/hooks.py:311]:
```python
# manual option 1
loss = ...
loss.backward()
...

# option 2 (automates scaling, etc)
loss = ...
self.backward(loss, opt)
```

File: pytorch_lightning/plugins/apex.py (hunk 2 above)
Comment (williamFalcon) [reply pytorch_lightning/plugins/apex.py]: done

**Comment by SeanNaren** [comment]
LGTM, much cleaner and the abstraction reduces code complexity.

File: docs/source/optimizers.rst [hunk 3]
```diff
@@ -1,7 +1,88 @@
...
+------
+
+Manual optimization
+===================
+For advanced research topics like reinforcement learning, sparse coding, or GAN research, it may be desirable
+to manually manage the optimization process. To do so, do the following:
+
+* Ignore the optimizer_idx argument
+* So we can scale the loss automatically for you use self.backward(loss) instead of loss.backward()
+
```
Comment (ananthsub) [inline docs/source/optimizers.rst:26]: do the same caveats from before apply? can they be included in the docs?

File: docs/source/optimizers.rst [hunk 4]
```diff
@@ -1,7 +1,88 @@
...
+In the case of multiple optimizers, Lightning does the following:
+
+.. code-block:: python
+
+    for epoch in epochs:
+      for batch in data:
+         for opt in optimizers:
+            train_step(opt)
+            opt.step()
```
Comment (ananthsub) [inline docs/source/optimizers.rst:81]: should this also mention that lightning disables grads based on the optimizer index being passed to training_step?

File: docs/source/optimizers.rst [hunk 5]
```diff
@@ -1,7 +1,88 @@
...
+        loss_b.step()
+        loss_b.zero_grad()
+
+.. note:: This is only recommended for experts who need ultimate flexibility
+
+------
+
+Automatic optimization
+======================
+With Lightning 90% of users don't have to think about when to call .backward(), .step(), .zero_grad(), since
```
Comment (ananthsub) [inline docs/source/optimizers.rst:56]:
```suggestion
With Lightning, most of users don't have to think about when to call .backward(), .step(), .zero_grad(), since
```
unless you're sure about the 90% :)

File: docs/source/optimizers.rst (hunk 3 above)
Comment (williamFalcon) [reply docs/source/optimizers.rst:26]: which caveats?

**Comment by williamFalcon** [comment]
horovod is failing to install... trying on another branch

File: docs/source/optimizers.rst [hunk 6]
```diff
@@ -1,7 +1,89 @@
...
+For advanced research topics like reinforcement learning, sparse coding, or GAN research, it may be desirable
+to manually manage the optimization process. To do so, do the following:
+
+* Ignore the optimizer_idx argument
+* So we can scale the loss automatically for you use self.backward(loss) instead of loss.backward()
+
+.. code-block:: python
+
+    def training_step(self, batch, batch_idx, optimizer_idx):
+        # ignore optimizer_idx
```
Comment (ananthsub) [inline docs/source/optimizers.rst:30]: training_step is still called once per optimizer index right? So the training step would need to ignore optimizer_idx > 0?

File: docs/source/optimizers.rst [hunk 7]
```diff
@@ -1,7 +1,89 @@
...
+        opt_g.zero_grad()
+
+        # do anything you want
+        loss_b = ...
+
+        # pass in any args that loss.backward() normally takes
+        self.backward(loss_b, opt_d, retain_graph=True)
+        self.backward(loss_b, opt_d, retain_graph=True)
+        loss_b.step()
+        loss_b.zero_grad()
```
Comment (rohitgr7) [inline docs/source/optimizers.rst:48]: you mean?
```suggestion
        opt_d.step()
        opt_d.zero_grad()
```
